modules/040-node-manager/hooks/yc_delete_preemptible_instances.go

```diff
@@ -0,0 +1,365 @@
+		return nil
+	}
+
+	for _, nodeRaw := range input.Snapshots["nodes"] {
+		if nodeRaw == nil {
+			continue
+		}
+
+		node, ok := nodeRaw.(*Node)
+		if !ok {
+			return fmt.Errorf("failed to assert to *Node")
+		}
+
+		nodeNameToNodeMap[node.Name] = node
+	}
+
+	for _, ngStatusRaw := range input.Snapshots["nodegroupstatuses"] {
+		if ngStatusRaw == nil {
+			continue
+		}
+
+		ngStatus, ok := ngStatusRaw.(*NodeGroupStatus)
+		if !ok {
+			return fmt.Errorf("failed to assert to *NodeGroupStatus")
+		}
+
+		nodeGroupNameToNodeGroupStatus[ngStatus.Name] = ngStatus
+	}
+
+	for _, machineRaw := range input.Snapshots["machines"] {
+		machine, ok := machineRaw.(*Machine)
+		if !ok {
+			return fmt.Errorf("failed to assert to *Machine")
+		}
+
+		if machine.Terminating {
+			continue
+		}
+
+		if machine.MachineClassKind != "YandexMachineClass" {
+			continue
+		}
+
+		if !preemptibleMachineClassesSet.Has(machine.MachineClassName) {
+			continue
+		}
+
+		if node, ok := nodeNameToNodeMap[machine.Name]; ok {
+			machine.nodeCreationTimestamp = node.CreationTimestamp
+			machine.nodeGroup = node.NodeGroup
+		} else {
+			continue
+		}
+
+		machines = append(machines, machine)
+	}
+
+	if len(machines) == 0 {
+		return nil
+	}
+
+	for _, m := range getMachinesToDelete(timeNow, machines, nodeGroupNameToNodeGroupStatus) {
+		input.PatchCollector.Delete("machine.sapcloud.io/v1alpha1", "Machine", "d8-cloud-instance-manager", m)
+	}
+
+	return nil
+}
+
+func getMachinesToDelete(timeNow time.Time, machines []*Machine, ngToNgStatusMap map[string]*NodeGroupStatus) (machinesToDelete []string) {
+	sort.Slice(machines, func(i, j int) bool {
+		return machines[i].nodeCreationTimestamp.Before(&machines[j].nodeCreationTimestamp)
+	})
+
+	// we spread deletion over preemptibleDeletionPeriod partitioned by hookExecutionSchedule
+	// that is the calculated batch for Machines that are older than 20 hours, but younger than 24 hours
+	batch := len(machines) / int(preemptibleDeletionPeriod/hookExecutionSchedule)
+	if batch == 0 {
+		batch = 1
+	}
+
+	var criticallyOldMachinesToDelete []string
+	const durationThresholdForDeletion = preemptibleInstanceMaxLifetime - preemptibleDeletionPeriod
+
+	for _, currentMachine := range machines {
+		// skip young Machines
+		if currentMachine.nodeCreationTimestamp.Time.Add(durationThresholdForDeletion).After(timeNow) {
+			continue
+		}
+
+		// skip Machines in NodeGroups that violate NodeGroup readiness ratio
+		ngStatus, ok := ngToNgStatusMap[currentMachine.nodeGroup]
+		if !ok {
+			continue
+		}
+		if (float64(ngStatus.Ready) / float64(ngStatus.Nodes)) < nodeGroupReadinessRatio {
+			continue
+		}
+
+		if (len(machinesToDelete) < batch) && (currentMachine.nodeCreationTimestamp.Time.Add(preemptibleInstanceMaxLifetime).After(timeNow)) {
+			machinesToDelete = append(machinesToDelete, currentMachine.Name)
+		}
+
+		if currentMachine.nodeCreationTimestamp.Time.Add(preemptibleInstanceMaxLifetime).Before(timeNow) {
```

Comment:
1. Should these machines be deleted in the first place?
2. Why the batch logic is not applied to critically old machines as machines? We know that real-world instances can live more than 24 hours in yandex cloud. IMO they should be counted and limited like machines to delete.

Reply:
1. They should. Here is how I see it: let MCM gracefully drain and terminate them, instead of being forcefully (in 30 seconds) stopped by Yandex.Cloud's hypervisor.
2. We have to remove these Machines ASAP. MCM will take care of draining them, we can't allow Yandex.Cloud to gracelessly terminate them.

https://cloud.yandex.com/en-ru/docs/compute/concepts/preemptible-vm

Reply:
Rewrote implementation to delete 10% of old Machines every 15 minutes.